web-api/src/middleware/request_context.rs

```diff
@@ -0,0 +1,184 @@
+// Copyright 2022 Xayn AG
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, version 3.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use std::{future::Future, sync::Arc};
+
+use actix_web::{
+    body::BoxBody,
+    dev::{Service, ServiceRequest, ServiceResponse},
+    web::Data,
+    HttpMessage,
+    HttpRequest,
+};
+use anyhow::anyhow;
+use futures_util::{
+    future::{self, Either},
+    FutureExt,
+};
+use reqwest::header::HeaderValue;
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+use tracing::{info_span, instrument, trace, Instrument};
+use uuid::Uuid;
+
+use crate::{error::early_failure::middleware_failure, tenants};
+
+pub(crate) struct RequestContext {
+    #[allow(unused)]
+    pub(crate) tenant_id: TenantId,
+    pub(crate) request_id: RequestId,
```

Comment:
> which mode we are without having to refer back to the configuration.

There is no `mode`.

Any tenant including legacy ones have a id (in case of legacy tenants it's generated on the fly if it doesn't exist yet)..

> This will also enable us in separating the request_id part from the tenant part, since they are not really related.

They are deeply related as they are both part of the request context (we will pass the request id through a header instead of generating it in the future).  And both must be given when setting up the logging span.
Splitting it apart without any need for this split is IMHO just making things unnecessary complicated.

Reply:
We don't need tenants to have their id. The code for multi-tenant is fortunately very contained.
If we are in a multi-tenant we check for the header and use `SET ROLE` other we do nothing. 

> future deployments of single tenant setups will use the same template as our multi tenant setup, not doing so would add massive maintenance overhead.

No, the two infrastructures will run in parallel till we have the resource to merge them.  It will not incur any overhead because we are going to use what we already have for a single tenant and we are going to deploy one multi-tenant environment only for customers to test the platform, not production. 
Any work that involves trying to partially migrate the current tenant to a multi-tenant-like setup is a waste of resources. We don't need it for our current goals, it is a middle-term or long target that we can address in the future with dedicated stories.

Reply:
> No, the two infrastructures will run in parallel till we have the resource to merge them.  It will not incur in any overhead because we are going to use what we already have for single tenant.

The two are completely orthogonal to each other, they have been deeply connected in the implementation but one can leave without the other. Is not an optimization that is refactoring the code so that each part works only on one concept. I didn't mark it correctly and I'm also fine with leaving it as it is now.